### PR TITLE
Updated Workflow page to reflect the changes in the api.

### PR DESCRIPTION
**Pull request by sagarkapare**

### CODE REVIEW

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 1]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
 failure, the execution is stopped at the failed job and no subsequent jobs in the sequence
 are executed.
 
-To process MapReduce or Spark programs in sequence, specify
-``addWorkflow()`` in your application::
+To execute MapReduce or Spark programs in the Workflow, you will need to add them in your
+application along with the Workflow. You can optionally add schedule to the Workflow.::
```
Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: "add a schedule"

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 2]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+    addWorkflow(new MyWorkflow());
+    scheduleWorkflow("FiveHourSchedule", "0 */5 * * *", "MyWorkflow");
+    ...
+  }
 
-You'll then implement the ``Workflow`` interface, which requires the
-``configure()`` method. From within ``configure``, call the
-``addSchedule()`` method to run a WorkFlow job periodically::
+You'll then extend the ``AbstractWorkflow`` class and implement the
+``configure()`` method. From within ``configure``, you can add multiple MapReduce, Spark
```
Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: "Inside `configure` you can add ..."

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 3]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+or Custom Action programs to the Workflow. The programs will get executed in a sequence in
```
Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: "in the sequence..."

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 4]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
-If there is only one MapReduce or Spark program to be run as a part of a WorkFlow,
-use the ``onlyWith()`` method after ``setDescription()`` when building
-the Workflow::
+``MyWorkflow`` will be executed every 5 hours. During each execution of the Workflow - ``MyMapReduce``,
+``MySpark``, ``AnotherMapReduce`` and ``MyAction`` programs will be executed in a sequence.
+
+.. rubric::  Workflow Custom Action
 
-  public static class PurchaseHistoryWorkflow implements Workflow {
+Apart from MapReduce and Spark programs, Workflow can be extended to execute custom actions. For
```
Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: "... can also execute custom actions."

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 5]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+defining custom action, you will need to extend the ``AbstractWorkflowAction`` and implement the ``run()`` method.::
```
Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: "To define a custom action..."

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 6]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
 failure, the execution is stopped at the failed job and no subsequent jobs in the sequence
 are executed.
 
-To process MapReduce or Spark programs in sequence, specify
-``addWorkflow()`` in your application::
+To execute MapReduce or Spark programs in the Workflow, you will need to add them in your
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: To execute MapReduce or Spark programs in a Workflow,

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 2 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]:
````
Inside ``configure``, you can add multiple MapReduce, Spark, or Custom Action programs to the Workflow. The programs will be executed in the order they are specified in the ``configure`` method.
````

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 7]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+        addSpark("MySpark");
+        addMapReduce("AnotherMapReduce");
+        addAction(new MyAction());
     }
   }
 
-If there is only one MapReduce or Spark program to be run as a part of a WorkFlow,
-use the ``onlyWith()`` method after ``setDescription()`` when building
-the Workflow::
+``MyWorkflow`` will be executed every 5 hours. During each execution of the Workflow - ``MyMapReduce``,
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]:
````
In this example, the ``MyWorkflow`` will be executed every 5 hours. During each execution of the Workflow, the ``MyMapReduce``, ``MySpark``, ``AnotherMapReduce``, and ``MyAction`` programs will be executed in order.
````

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 4 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: In addition to MapReduce and Spark programs,…

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 8]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+        setDescription("MyWorkflow description");
+        addMapReduce("MyMapReduce");
+        addSpark("MySpark");
+        addMapReduce("AnotherMapReduce");
+        addAction(new MyAction());
     }
   }
 
-If there is only one MapReduce or Spark program to be run as a part of a WorkFlow,
-use the ``onlyWith()`` method after ``setDescription()`` when building
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Has the ``onlyWith`` method gone away?

Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Yes.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 5 above)
Comment (gokulavasan) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Do we talk about what custom action is anywhere in the documentation? If not we need to add that in.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 9]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
     @Override
-    public WorkflowSpecification configure() {
-      return WorkflowSpecification.Builder.with() .setName("PurchaseHistoryWorkflow")
-        .setDescription("PurchaseHistoryWorkflow description")
-        .onlyWith(new PurchaseHistoryBuilder())
-        .addSchedule(new DefaultSchedule("FiveMinuteSchedule", "Run every 5 minutes",
-                     "0/5 * * * *", Schedule.Action.START))
-        .build();
+    public void run() {
+      // your code goes here
```
Comment (gokulavasan) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Would it be nice to have an example what can go in here? That might be part of custom action documentation.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 5 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: To define a custom action, you will need to extend ``AbstractWorkflowAction`` and implement its ``run`` method::

\[note that there is no period before the double-colon]

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 10]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
     }
   }
 
+The custom action then can be added to the Workflow using ``addAction()`` method as shown above.
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: using the ``addAction`` method as shown above.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 11]
```diff
@@ -16,49 +16,55 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+application along with the Workflow. You can optionally add schedule to the Workflow.::
 
   public void configure() {
     ...
-    addWorkflow(new PurchaseHistoryWorkflow());
+    addMapReduce(new MyMapReduce());
+    addMapReduce(new AnotherMapReduce());
+    addSpark(new MySpark());
+    addWorkflow(new MyWorkflow());
+    scheduleWorkflow("FiveHourSchedule", "0 */5 * * *", "MyWorkflow");
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Do we document how to set up these schedules, and what the syntax is?

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 1 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 2 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 4 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Changed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 5 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: done.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 6 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Changed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 7 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 10 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Fixed.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 11 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Yes. We have that as a part of java doc.

Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: What about adding a link to that here? I can rebuild the Javadocs and send you a link.

Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Ok sure.

**Comment by gokulavasan** [comment]
Changes are fine. But I hope we will address the WorkflowAction documentation as mentioned in this JIRA : https://issues.cask.co/browse/CDAP-1128

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 12]
```diff
@@ -16,49 +16,56 @@ transferred to the next job in sequence until the last job in the sequence is ex
 failure, the execution is stopped at the failed job and no subsequent jobs in the sequence
 are executed.
 
-To process MapReduce or Spark programs in sequence, specify
-``addWorkflow()`` in your application::
+To execute MapReduce or Spark programs in a Workflow, you will need to add them in your
+application along with the Workflow. You can optionally add a schedule to the Workflow::
```
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: You can optionally add a schedule (such as a `crontab schedule <../../reference-manual/javadocs/co/cask/cdap/api/app/AbstractApplication.html#scheduleWorkflow(java.lang.String,%20java.lang.String,%20java.lang.String)>`__) to the Workflow::

**Comment by sagarkapare** [comment]
@awholegunch Addressed comments. Please take a look again. Thanks!

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 13]
```diff
@@ -16,49 +16,58 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
-the Workflow::
+In this example, the ``MyWorkflow`` will be executed every 5 hours. During each execution of the Workflow,
+the ``MyMapReduce``, ``MySpark``, ``AnotherMapReduce``, and ``MyAction`` programs will be executed in order.
+
+.. rubric::  Workflow Custom Action
 
-  public static class PurchaseHistoryWorkflow implements Workflow {
+In addition to MapReduce and Spark programs, Workflow can also execute custom actions. Custom actions are
+implemented in Java and can perform tasks such as make a REST call to an external system. To define
+custom action, you will need to extend the ``AbstractWorkflowAction`` and implement the ``run()`` method::
```
Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: "a custom action"?

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 14]
```diff
@@ -16,49 +16,58 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
-use the ``onlyWith()`` method after ``setDescription()`` when building
-the Workflow::
+In this example, the ``MyWorkflow`` will be executed every 5 hours. During each execution of the Workflow,
+the ``MyMapReduce``, ``MySpark``, ``AnotherMapReduce``, and ``MyAction`` programs will be executed in order.
+
+.. rubric::  Workflow Custom Action
 
-  public static class PurchaseHistoryWorkflow implements Workflow {
+In addition to MapReduce and Spark programs, Workflow can also execute custom actions. Custom actions are
+implemented in Java and can perform tasks such as make a REST call to an external system. To define
```
Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: I'm not a fan of giving this as an example. It brings up the obvious question of whether you can make REST calls to `Service`s and not only to external systems. Taking a quick look at it, and it doesn't seem that `WorkflowAction` supports ServiceDiscovery (any idea why? Should it?).

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 15]
```diff
@@ -16,49 +16,58 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+    scheduleWorkflow("FiveHourSchedule", "0 */5 * * *", "MyWorkflow");
+    ...
+  }
 
-You'll then implement the ``Workflow`` interface, which requires the
-``configure()`` method. From within ``configure``, call the
-``addSchedule()`` method to run a WorkFlow job periodically::
+You'll then extend the ``AbstractWorkflow`` class and implement the
+``configure()`` method. Inside ``configure``, you can add multiple MapReduce, Spark, or Custom Action
+programs to the Workflow. The programs will be executed in the order they are specified in the ``configure``
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: or custom actions to the workflow. (A custom action is not a program)

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst [hunk 16]
```diff
@@ -16,49 +16,58 @@ transferred to the next job in sequence until the last job in the sequence is ex
...
+        addAction(new MyAction());
     }
   }
 
-If there is only one MapReduce or Spark program to be run as a part of a WorkFlow,
-use the ``onlyWith()`` method after ``setDescription()`` when building
-the Workflow::
+In this example, the ``MyWorkflow`` will be executed every 5 hours. During each execution of the Workflow,
+the ``MyMapReduce``, ``MySpark``, ``AnotherMapReduce``, and ``MyAction`` programs will be executed in order.
+
```
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: similarly here, MyAction is not a program.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 14 above)
Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: agreed. It should, but let's do that in a later release.

Comment (anew) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: How about changing this to, sending an email?

Comment (ParthGandhi) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: +1

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 16 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: the ``MyMapReduce``, ``MySpark``, and ``AnotherMapReduce`` programs and the ``MyAction`` custom action will be executed in order.

Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Addressed comment.

File: cdap-docs/developers-manual/source/building-blocks/workflows.rst (hunk 15 above)
Comment (sagarkapare) [inline cdap-docs/developers-manual/source/building-blocks/workflows.rst]: Updated.

**Comment by awholegunch** [comment]
Please update the releases notes in the wiki to reflect this work.

**Comment by sagarkapare** [comment]
Wiki is updated with the draft release notes.

**Comment by anew** [comment]
LGTM

**Pull request merged**